### ISSUE TITLE
SplitLayer: Beam search not working

### ISSUE THREAD

**Issue by Zettelkasten** [rwth-i6/returnn#450]
Inside a `RecLayer`, I am accessing the output `split/0` of a `SplitLayer`:
```python
"split": {"class": "split", "size_splits": (5, 5), "axis": "F", "from": ["embed"]},
"output_prob": {"class": "softmax", "from": ["split/0", "base:encoder"], "target": "classes", "loss": "ce"},
```

During search, the construction of `output_prob` fails:
```
  File "/home/runner/work/returnn/returnn/returnn/tf/network.py", line 657, in _create_layer_layer_desc
    line: layer_desc = SearchChoices.translate_to_common_search_beam(layer_desc)
    locals:
      layer_desc = <local> {'target': 'classes', 'sources': [<InternalLayer output/'split/0' out_type=Data(shape=(5,), time_dim_axis=None, beam=SearchBeam(name='output/prev:output', beam_size=4), batch_shape_meta=[B,F|5])>, <LinearLayer 'encoder' out_type=Data(shape=(None, 5), batch_shape_meta=[B,T|'time:var:extern_data:da...
      SearchChoices = <local> <class 'returnn.tf.layers.base.SearchChoices'>
      SearchChoices.translate_to_common_search_beam = <local> <bound method SearchChoices.translate_to_common_search_beam of <class 'returnn.tf.layers.base.SearchChoices'>>
  File "/home/runner/work/returnn/returnn/returnn/tf/layers/base.py", line 1988, in translate_to_common_search_beam
    line: assert search_choices, "layer %r has beam %r but no search choices; from layer desc\n%s" % (
            layer, layer.output.beam, pformat(layer_desc))
    locals:
      search_choices = <local> None
      layer = <local> <InternalLayer output/'split/0' out_type=Data(shape=(5,), time_dim_axis=None, beam=SearchBeam(name='output/prev:output', beam_size=4), batch_shape_meta=[B,F|5])>
      layer.output = <local> Data(name='split/0_output', shape=(5,), time_dim_axis=None, beam=SearchBeam(name='output/prev:output', beam_size=4), batch_shape_meta=[B,F|5])
      layer.output.beam = <local> SearchBeam(name='output/prev:output', beam_size=4)
      pformat = <local> <function pformat at 0x7f9baaeafb90>
      layer_desc = <local> {'target': 'classes', 'sources': [<InternalLayer output/'split/0' out_type=Data(shape=(5,), time_dim_axis=None, beam=SearchBeam(name='output/prev:output', beam_size=4), batch_shape_meta=[B,F|5])>, <LinearLayer 'encoder' out_type=Data(shape=(None, 5), batch_shape_meta=[B,T|'time:var:extern_data:da...
AssertionError: layer <InternalLayer output/'split/0' out_type=Data(shape=(5,), time_dim_axis=None, beam=SearchBeam(name='output/prev:output', beam_size=4), batch_shape_meta=[B,F|5])> has beam SearchBeam(name='output/prev:output', beam_size=4) but no search choices; from layer desc
```

I added a test case in #449 and you can [find the full stack trace here](https://github.com/rwth-i6/returnn/runs/1945919540).

**Comment by albertz** [comment]
The sublayers of `SplitLayer` might not have the dependency correctly set.

**Comment by patrick-wilken** [comment]
Yes, the search choices are determined by calling `get_dep_layers()` recursively, which for `InternalLayer` (=`LayerBase`) looks at `self.sources`. But the `InternalLayer` created in `SplitLayer._make_split_layer()` has no sources.
So the first step to solve this is to make the parent layer a "dep_layer" of the sub-layer.
Currently, there is no "extra dependency" parameter for `LayerBase` (other layers have that), so you might want to try to add the parent layer as a (unused) source for now, which in theory should get you the beam:
```
return InternalLayer(name="%s/%i" % (self.name, idx), network=self.network, output=out, sources=[self])
```
This seems to do the right thing, however, the template for the layer is still missing the beam. Which may or may not be related to the changes I made here: https://github.com/rwth-i6/returnn/pull/398/commits/7b4e1ec5ce364eb7d61d094c4004664837ff7d84 I'm trying to figure it out...

**Comment by albertz** [comment]
Yep, I'm also wondering whether your (@patrick-wilken) changes are introducing any problems here...

But we could easily fix the template to include the beam (and should anyway if it is wrong). If that is the only remaining problem, then this should not really be a problem, and all should be fine.

**Comment by patrick-wilken** [comment]
See #453

**Comment by albertz** [comment]
I just merged #453. Is this fixed now?

**Comment by albertz** [comment]
I guess this can be closed now.
